In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [57]:
%%sql
WITH genre_counts AS (
SELECT g.name,ar.name artist, CAST(count(t.track_id) as float) counts
FROM customer c
JOIN invoice i ON c.customer_id = i.customer_id
JOIN invoice_line il ON i.invoice_id = il.invoice_id
JOIN track t ON il.track_id = t.track_id
JOIN genre g ON t.genre_id = g.genre_id
JOIN album a ON t.album_id = a.album_id
JOIN artist ar ON a.artist_id = ar.artist_id
WHERE c.country = 'USA'
GROUP BY 1,2)
 
SELECT genre_counts.name, genre_counts.artist, CAST (genre_counts.counts as int) counts,
ROUND(CAST (genre_counts.counts/(
SELECT CAST (count(t.track_id) as float)total
FROM customer c
JOIN invoice i ON c.customer_id = i.customer_id
JOIN invoice_line il ON i.invoice_id = il.invoice_id
JOIN track t ON il.track_id = t.track_id
JOIN genre g ON t.genre_id = g.genre_id
WHERE c.country = 'USA'
) as FLOAT)*100,2) percentage
FROM genre_counts
ORDER BY counts desc




Done.


[('Rock', 'Van Halen', 43, 4.09),
 ('Rock', 'The Rolling Stones', 37, 3.52),
 ('Rock', 'Nirvana', 35, 3.33),
 ('Alternative & Punk', 'Green Day', 32, 3.04),
 ('Alternative & Punk', 'R.E.M.', 32, 3.04),
 ('Blues', 'Eric Clapton', 31, 2.95),
 ('Rock', 'AC/DC', 29, 2.76),
 ('Rock', 'Pearl Jam', 29, 2.76),
 ('Rock', 'Jimi Hendrix', 28, 2.66),
 ('Rock', 'The Doors', 27, 2.57),
 ('R&B/Soul', 'Marvin Gaye', 26, 2.47),
 ('Metal', 'Black Sabbath', 24, 2.28),
 ('Rock', "Guns N' Roses", 23, 2.19),
 ('Metal', 'System Of A Down', 22, 2.09),
 ('Rock', 'Alice In Chains', 21, 2.0),
 ('Rock', 'Foo Fighters', 21, 2.0),
 ('Rock', 'Scorpions', 21, 2.0),
 ('Hip Hop/Rap', 'House Of Pain', 20, 1.9),
 ('Metal', 'Godsmack', 20, 1.9),
 ('Alternative', 'Chris Cornell', 19, 1.81),
 ('Alternative & Punk', 'Os Mutantes', 19, 1.81),
 ('Rock', 'Queen', 19, 1.81),
 ('Rock', 'Led Zeppelin', 17, 1.62),
 ('R&B/Soul', 'Amy Winehouse', 16, 1.52),
 ('Rock', 'The Who', 16, 1.52),
 ('Rock', 'Red Hot Chili Peppers', 15, 1.43),
 ('Rock', 'U2', 15, 1.43),
 ('Pop', 'Amy Winehouse', 14, 1.33),
 ('Alternative & Punk', 'Foo Fighters', 13, 1.24),
 ('Alternative & Punk', 'The Clash', 13, 1.24),
 ('Easy Listening', 'Frank Sinatra', 13, 1.24),
 ('Rock', 'Skank', 13, 1.24),
 ('Rock', 'Soundgarden', 13, 1.24),
 ('Metal', 'Metallica', 12, 1.14),
 ('Rock', 'Creedence Clearwater Revival', 12, 1.14),
 ('Rock', 'Deep Purple', 12, 1.14),
 ('Rock', 'Def Leppard', 10, 0.95),
 ('Rock', 'Pink Floyd', 10, 0.95),
 ('Alternative & Punk', 'JET', 9, 0.86),
 ('Jazz', 'Miles Davis', 9, 0.86),
 ('Metal', "Guns N' Roses", 9, 0.86),
 ('R&B/Soul', 'James Brown', 9, 0.86),
 ('Rock', 'The Police', 9, 0.86),
 ('Rock', 'Velvet Revolver', 9, 0.86),
 ('Metal', 'Apocalyptica', 8, 0.76),
 ('Metal', 'Judas Priest', 8, 0.76),
 ('Pop', 'U2', 8, 0.76),
 ('Rock', 'Alanis Morissette', 8, 0.76),
 ('Alternative', 'Audioslave', 7, 0.67),
 ('Rock', 'Aerosmith', 7, 0.67),
 ('Rock', 'Jamiroquai', 7, 0.67),
 ('Alternative', 'Cake', 6, 0.57),
 ('Reggae', 'UB40', 6, 0.57),
 ('Rock', 'R.E.M.', 6, 0.57),
 ('Rock', 'Stone Temple Pilots', 6, 0.57),
 ('Metal', 'Motörhead', 5, 0.48),
 ('Metal', 'Mötley Crüe', 5, 0.48),
 ('Rock', 'Rush', 5, 0.48),
 ('Electronica/Dance', 'Jamiroquai', 4, 0.38),
 ('Rock', 'Kiss', 4, 0.38),
 ('Rock', 'Lenny Kravitz', 4, 0.38),
 ('Rock', 'Marillion', 4, 0.38),
 ('Rock', 'Men At Work', 4, 0.38),
 ('Rock', 'Ozzy Osbourne', 4, 0.38),
 ('Rock', 'Santana', 4, 0.38),
 ('Alternative & Punk', 'Raimundos', 3, 0.29),
 ('Blues', 'Buddy Guy', 3, 0.29),
 ('Heavy Metal', 'Iron Maiden', 3, 0.29),
 ('Latin', 'Cássia Eller', 3, 0.29),
 ('Latin', 'Eric Clapton', 3, 0.29),
 ('Latin', 'Marisa Monte', 3, 0.29),
 ('Metal', 'Iron Maiden', 3, 0.29),
 ('Metal', 'Lenny Kravitz', 3, 0.29),
 ('Metal', 'Ozzy Osbourne', 3, 0.29),
 ('Rock', 'Raul Seixas', 3, 0.29),
 ('Alternative', 'Temple of the Dog', 2, 0.19),
 ('Alternative & Punk', 'Faith No More', 2, 0.19),
 ('Alternative & Punk', 'Pearl Jam', 2, 0.19),
 ('Alternative & Punk', 'Red Hot Chili Peppers', 2, 0.19),
 ('Blues', 'The Black Crowes', 2, 0.19),
 ('Jazz', 'Billy Cobham', 2, 0.19),
 ('Latin', 'Caetano Veloso', 2, 0.19),
 ('Latin', 'Legião Urbana', 2, 0.19),
 ('R&B/Soul', 'Jamiroquai', 2, 0.19),
 ('Rock', 'Accept', 2, 0.19),
 ('Rock', 'Audioslave', 2, 0.19),
 ('Rock', 'Iron Maiden', 2, 0.19),
 ('Rock', 'Joe Satriani', 2, 0.19),
 ('Rock', 'The Cult', 2, 0.19),
 ('Alternative', 'Calexico', 1, 0.1),
 ('Alternative & Punk', 'Audioslave', 1, 0.1),
 ('Alternative & Punk', 'Body Count', 1, 0.1),
 ('Alternative & Punk', 'Titãs', 1, 0.1),
 ('Classical', 'Luciano Pavarotti', 1, 0.1),
 ('Classical', 'Maurizio Pollini', 1, 0.1),
 ('Classical', "The King's Singers", 1, 0.1),
 ('Classical', 'Yo-Yo Ma', 1, 0.1),
 ('Electronica/Dance', 'Karsh Kale', 1, 0.1),
 ('Jazz', 'Aisha Duo', 1, 0.1),
 ('Jazz', 'Incognito', 1, 0.1),
 ('Jazz', 'Spyro Gyra', 1, 0.1),
 ('Latin', 'Antônio Carlos Jobim', 1, 0.1),
 ('Latin', 'Ed Motta', 1, 0.1),
 ('Latin', 'Funk Como Le Gusta', 1, 0.1),
 ('Latin', 'Jorge 

# Recomendation from three best sold albums
Rock 	Van Halen 	43 	4.09
Rock 	The Rolling Stones 	37 	3.52
Rock 	Nirvana 	35 	3.33


In [16]:
%%sql
SELECT count(t.track_id) total
FROM customer c
JOIN invoice i ON c.customer_id = i.customer_id
JOIN invoice_line il ON i.invoice_id = il.invoice_id
JOIN track t ON il.track_id = t.track_id
JOIN genre g ON t.genre_id = g.genre_id
WHERE c.country = 'USA'

Done.


[(1051,)]

In [17]:
%%sql
SELECT g.name, count(t.track_id) counts
FROM customer c
JOIN invoice i ON c.customer_id = i.customer_id
JOIN invoice_line il ON i.invoice_id = il.invoice_id
JOIN track t ON il.track_id = t.track_id
JOIN genre g ON t.genre_id = g.genre_id
WHERE c.country = 'USA'
GROUP BY 1

Done.


[('Alternative', 35),
 ('Alternative & Punk', 130),
 ('Blues', 36),
 ('Classical', 4),
 ('Easy Listening', 13),
 ('Electronica/Dance', 5),
 ('Heavy Metal', 3),
 ('Hip Hop/Rap', 20),
 ('Jazz', 14),
 ('Latin', 22),
 ('Metal', 124),
 ('Pop', 22),
 ('R&B/Soul', 53),
 ('Reggae', 6),
 ('Rock', 561),
 ('Soundtrack', 2),
 ('TV Shows', 1)]

# Write a query that finds the total dollar amount of sales assigned to each sales support agent within the company. Add any extra attributes for that employee that you find are relevant to the analysis

In [66]:
%%sql
WITH genre_counts AS (
SELECT e.first_name||" "||e.last_name as full_name,
    c.company customer_name,c.city customer_city, ROUND (CAST(sum(i.total) as float),2) amount
FROM employee e
JOIN customer c ON e.employee_id = c.support_rep_id
JOIN invoice i ON c.customer_id = i.customer_id
GROUP BY 1,2,3)

select * from genre_counts
order by 4 desc

Done.


[('Jane Peacock', None, 'London', 166.32),
 ('Margaret Park', None, 'Paris', 151.47),
 ('Margaret Park', 'JetBrains s.r.o.', 'Prague', 144.54),
 ('Steve Johnson', None, 'Prague', 128.7),
 ('Jane Peacock', None, 'Dublin', 114.84),
 ('Jane Peacock', None, 'Delhi', 111.87),
 ('Jane Peacock', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'São José dos Campos', 108.9),
 ('Margaret Park', None, 'Brasília', 106.92),
 ('Margaret Park', None, 'Lisbon', 102.96),
 ('Jane Peacock', None, 'Bordeaux', 99.99),
 ('Jane Peacock', None, 'Montréal', 99.99),
 ('Steve Johnson', None, 'Madrid', 98.01),
 ('Steve Johnson', 'Microsoft Corporation', 'Redmond', 98.01),
 ('Steve Johnson', None, 'Santiago', 97.02),
 ('Margaret Park', None, 'Mountain View', 95.04),
 ('Jane Peacock', None, 'Frankfurt', 94.05),
 ('Margaret Park', None, 'Orlando', 92.07),
 ('Jane Peacock', None, 'Ottawa', 91.08),
 ('Steve Johnson', None, 'Reno', 91.08),
 ('Margaret Park', None, 'Fort Worth', 86.13),
 ('Steve Johnson', None, 'Berlin', 85.14),
 ('Margaret Park', None, 'Tucson', 84.15),
 ('Jane Peacock', 'Riotur', 'Rio de Janeiro', 82.17),
 ('Margaret Park', None, 'Porto', 82.17),
 ('Steve Johnson', None, 'Stuttgart', 82.17),
 ('Margaret Park', None, 'Sidney', 81.18),
 ('Jane Peacock', None, 'Helsinki', 79.2),
 ('Jane Peacock', None, 'New York', 79.2),
 ('Steve Johnson', None, 'Edinburgh ', 79.2),
 ('Jane Peacock', None, 'Budapest', 78.21),
 ('Margaret Park', None, 'Warsaw', 76.23),
 ('Steve Johnson', None, 'Madison', 76.23),
 ('Jane Peacock', None, 'Yellowknife', 75.24),
 ('Steve Johnson', None, 'Stockholm', 75.24),
 ('Margaret Park', 'Google Inc.', 'Mountain View', 74.25),
 ('Jane Peacock', None, 'Berlin', 73.26),
 ('Jane Peacock', None, 'Dijon', 73.26),
 ('Margaret Park', None, 'Oslo', 72.27),
 ('Steve Johnson', None, 'Salt Lake City', 72.27),
 ('Jane Peacock', None, 'Bangalore', 71.28),
 ('Jane Peacock', None, 'Chicago', 71.28),
 ('Margaret Park', None, 'Winnipeg', 70.29),
 ('Steve Johnson', None, 'Vienne', 69.3),
 ('Steve Johnson', 'Banco do Brasil S.A.', 'São Paulo', 69.3),
 ('Jane Peacock', 'Rogers Canada', 'Vancouver', 66.33),
 ('Margaret Park', None, 'Boston', 66.33),
 ('Steve Johnson', None, 'Amsterdam', 65.34),
 ('Steve Johnson', None, 'Lyon', 64.35),
 ('Steve Johnson', None, 'Halifax', 62.37),
 ('Margaret Park', None, 'Brussels', 60.39),
 ('Margaret Park', 'Woodstock Discos', 'São Paulo', 60.39),
 ('Jane Peacock', 'Apple Inc.', 'Cupertino', 54.45),
 ('Steve Johnson', None, 'Rome', 50.49),
 ('Jane Peacock', None, 'Toronto', 40.59),
 ('Margaret Park', None, 'Buenos Aires', 39.6),
 ('Margaret Park', None, 'Copenhagen', 37.62),
 ('Steve Johnson', 'Telus', 'Edmonton', 29.7)]

# Most of sales is done on no name customers

# Write a query that collates data on purchases from different countries.

    Where a country has only one customer, collect them into an "Other" group.
    The results should be sorted by the total sales from highest to lowest, with the "Other" group at the very bottom.
    For each country, include:
        total number of customers
        total value of sales
        average value of sales per customer
        average order value


In [71]:
%%sql
WITH genre_counts AS (
SELECT c.country, count(c.customer_id), ROUND(sum(i.total),2), ROUND(avg(i.total),2)
FROM customer c
JOIN invoice i ON c.customer_id = i.customer_id
JOIN invoice_line il ON i.invoice_id = il.invoice_id
GROUP BY 1)

select * from genre_counts
order by 2 desc

Done.


[('USA', 1051, 10405.89, 9.9),
 ('Canada', 541, 5489.55, 10.15),
 ('Brazil', 432, 4059.0, 9.4),
 ('France', 393, 3972.87, 10.11),
 ('Germany', 338, 3441.24, 10.18),
 ('Czech Republic', 276, 3183.84, 11.54),
 ('United Kingdom', 248, 2498.76, 10.08),
 ('Portugal', 187, 1755.27, 9.39),
 ('India', 185, 1887.93, 10.21),
 ('Ireland', 116, 1433.52, 12.36),
 ('Spain', 99, 1076.13, 10.87),
 ('Chile', 98, 912.78, 9.31),
 ('Australia', 82, 940.5, 11.47),
 ('Finland', 80, 685.08, 8.56),
 ('Hungary', 79, 830.61, 10.51),
 ('Poland', 77, 690.03, 8.96),
 ('Sweden', 76, 754.38, 9.93),
 ('Norway', 73, 664.29, 9.1),
 ('Austria', 70, 649.44, 9.28),
 ('Netherlands', 66, 544.5, 8.25),
 ('Belgium', 61, 567.27, 9.3),
 ('Italy', 51, 468.27, 9.18),
 ('Argentina', 40, 396.0, 9.9),
 ('Denmark', 38, 196.02, 5.16)]

# Write a query that categorizes each invoice as either an album purchase or not, and calculates the following summary statistics:

    Number of invoices
    Percentage of invoices


In [74]:
%%sql
select invoice_id, count(invoice_id) from invoice_line
group by invoice_id

Done.


[(1, 16),
 (2, 10),
 (3, 2),
 (4, 8),
 (5, 17),
 (6, 2),
 (7, 11),
 (8, 10),
 (9, 9),
 (10, 2),
 (11, 11),
 (12, 4),
 (13, 1),
 (14, 1),
 (15, 4),
 (16, 9),
 (17, 11),
 (18, 5),
 (19, 7),
 (20, 10),
 (21, 9),
 (22, 2),
 (23, 10),
 (24, 12),
 (25, 8),
 (26, 2),
 (27, 8),
 (28, 3),
 (29, 6),
 (30, 11),
 (31, 20),
 (32, 13),
 (33, 4),
 (34, 1),
 (35, 5),
 (36, 7),
 (37, 10),
 (38, 7),
 (39, 8),
 (40, 10),
 (41, 8),
 (42, 9),
 (43, 7),
 (44, 12),
 (45, 8),
 (46, 11),
 (47, 6),
 (48, 4),
 (49, 5),
 (50, 7),
 (51, 8),
 (52, 12),
 (53, 3),
 (54, 18),
 (55, 10),
 (56, 17),
 (57, 1),
 (58, 6),
 (59, 11),
 (60, 8),
 (61, 12),
 (62, 3),
 (63, 6),
 (64, 13),
 (65, 2),
 (66, 7),
 (67, 11),
 (68, 11),
 (69, 14),
 (70, 1),
 (71, 17),
 (72, 3),
 (73, 15),
 (74, 2),
 (75, 4),
 (76, 13),
 (77, 6),
 (78, 9),
 (79, 14),
 (80, 1),
 (81, 6),
 (82, 7),
 (83, 4),
 (84, 12),
 (85, 5),
 (86, 1),
 (87, 8),
 (88, 5),
 (89, 2),
 (90, 11),
 (91, 11),
 (92, 20),
 (93, 10),
 (94, 8),
 (95, 11),
 (96, 6),
 (97, 6),
 (98, 10),
 (99, 18),
 (100, 6),
 (101, 4),
 (102, 10),
 (103, 6),
 (104, 2),
 (105, 8),
 (106, 8),
 (107, 11),
 (108, 5),
 (109, 11),
 (110, 6),
 (111, 5),
 (112, 17),
 (113, 10),
 (114, 17),
 (115, 9),
 (116, 11),
 (117, 11),
 (118, 1),
 (119, 2),
 (120, 6),
 (121, 9),
 (122, 5),
 (123, 12),
 (124, 6),
 (125, 10),
 (126, 4),
 (127, 11),
 (128, 7),
 (129, 11),
 (130, 4),
 (131, 3),
 (132, 2),
 (133, 10),
 (134, 8),
 (135, 3),
 (136, 4),
 (137, 2),
 (138, 11),
 (139, 5),
 (140, 8),
 (141, 18),
 (142, 15),
 (143, 5),
 (144, 3),
 (145, 5),
 (146, 13),
 (147, 5),
 (148, 9),
 (149, 9),
 (150, 3),
 (151, 3),
 (152, 6),
 (153, 14),
 (154, 3),
 (155, 9),
 (156, 11),
 (157, 1),
 (158, 6),
 (159, 6),
 (160, 9),
 (161, 8),
 (162, 9),
 (163, 9),
 (164, 4),
 (165, 8),
 (166, 10),
 (167, 6),
 (168, 6),
 (169, 9),
 (170, 10),
 (171, 13),
 (172, 8),
 (173, 11),
 (174, 7),
 (175, 18),
 (176, 7),
 (177, 2),
 (178, 3),
 (179, 10),
 (180, 10),
 (181, 7),
 (182, 6),
 (183, 24),
 (184, 1),
 (185, 9),
 (186, 3),
 (187, 7),
 (188, 11),
 (189, 4),
 (190, 5),
 (191, 4),
 (192, 9),
 (193, 2),
 (194, 1),
 (195, 11),
 (196, 10),
 (197, 11),
 (198, 11),
 (199, 3),
 (200, 8),
 (201, 12),
 (202, 2),
 (203, 10),
 (204, 11),
 (205, 10),
 (206, 5),
 (207, 4),
 (208, 6),
 (209, 4),
 (210, 11),
 (211, 9),
 (212, 12),
 (213, 11),
 (214, 5),
 (215, 2),
 (216, 5),
 (217, 10),
 (218, 9),
 (219, 3),
 (220, 9),
 (221, 8),
 (222, 2),
 (223, 18),
 (224, 10),
 (225, 6),
 (226, 11),
 (227, 10),
 (228, 4),
 (229, 3),
 (230, 8),
 (231, 3),
 (232, 4),
 (233, 6),
 (234, 1),
 (235, 7),
 (236, 3),
 (237, 16),
 (238, 10),
 (239, 16),
 (240, 3),
 (241, 2),
 (242, 10),
 (243, 11),
 (244, 5),
 (245, 3),
 (246, 9),
 (247, 11),
 (248, 7),
 (249, 2),
 (250, 11),
 (251, 7),
 (252, 8),
 (253, 12),
 (254, 10),
 (255, 4),
 (256, 5),
 (257, 11),
 (258, 1),
 (259, 7),
 (260, 8),
 (261, 9),
 (262, 8),
 (263, 5),
 (264, 1),
 (265, 7),
 (266, 6),
 (267, 2),
 (268, 11),
 (269, 1),
 (270, 11),
 (271, 11),
 (272, 9),
 (273, 17),
 (274, 5),
 (275, 8),
 (276, 19),
 (277, 5),
 (278, 3),
 (279, 5),
 (280, 11),
 (281, 14),
 (282, 17),
 (283, 9),
 (284, 11),
 (285, 10),
 (286, 2),
 (287, 2),
 (288, 11),
 (289, 7),
 (290, 5),
 (291, 3),
 (292, 10),
 (293, 5),
 (294, 11),
 (295, 1),
 (296, 13),
 (297, 3),
 (298, 9),
 (299, 6),
 (300, 12),
 (301, 8),
 (302, 4),
 (303, 10),
 (304, 7),
 (305, 15),
 (306, 11),
 (307, 10),
 (308, 11),
 (309, 9),
 (310, 9),
 (311, 6),
 (312, 10),
 (313, 5),
 (314, 10),
 (315, 5),
 (316, 2),
 (317, 12),
 (318, 8),
 (319, 10),
 (320, 17),
 (321, 1),
 (322, 10),
 (323, 8),
 (324, 5),
 (325, 10),
 (326, 9),
 (327, 9),
 (328, 4),
 (329, 1),
 (330, 13),
 (331, 6),
 (332, 10),
 (333, 6),
 (334, 15),
 (335, 15),
 (336, 16),
 (337, 10),
 (338, 5),
 (339, 8),
 (340, 3),
 (341, 12),
 (342, 3),
 (343, 6),
 (344, 17),
 (345, 10),
 (346, 6),
 (347, 4),
 (348, 9),
 (349, 1),
 (350, 3),
 (351, 10),
 (352, 1),
 (353, 10),
 (354, 5),
 (355, 7),
 (356, 7),
 (357, 3),
 (358, 7),
 (359, 9),
 (360, 4),
 (361, 2),
 (362,

In [77]:
%%sql
select invoice_id, album_id, count(invoice_id) from  invoice_line il
JOIN track t ON il.track_id = t.track_id 
group by 1,2

Done.


[(1, 91, 16),
 (2, 20, 1),
 (2, 34, 1),
 (2, 39, 1),
 (2, 66, 1),
 (2, 73, 1),
 (2, 134, 1),
 (2, 190, 1),
 (2, 215, 1),
 (2, 218, 1),
 (2, 322, 1),
 (3, 203, 1),
 (3, 214, 1),
 (4, 58, 1),
 (4, 67, 1),
 (4, 146, 1),
 (4, 153, 1),
 (4, 203, 1),
 (4, 207, 1),
 (4, 260, 1),
 (4, 314, 1),
 (5, 163, 17),
 (6, 5, 1),
 (6, 83, 1),
 (7, 6, 1),
 (7, 55, 1),
 (7, 67, 1),
 (7, 87, 1),
 (7, 106, 1),
 (7, 163, 1),
 (7, 211, 1),
 (7, 214, 1),
 (7, 239, 1),
 (7, 305, 1),
 (7, 322, 1),
 (8, 9, 1),
 (8, 46, 1),
 (8, 118, 1),
 (8, 125, 1),
 (8, 144, 1),
 (8, 145, 1),
 (8, 183, 1),
 (8, 185, 1),
 (8, 203, 1),
 (8, 233, 1),
 (9, 20, 1),
 (9, 21, 1),
 (9, 79, 1),
 (9, 83, 1),
 (9, 101, 1),
 (9, 162, 1),
 (9, 173, 1),
 (9, 271, 1),
 (9, 321, 1),
 (10, 215, 1),
 (10, 257, 1),
 (11, 49, 1),
 (11, 89, 1),
 (11, 115, 1),
 (11, 181, 1),
 (11, 185, 1),
 (11, 195, 1),
 (11, 197, 1),
 (11, 211, 1),
 (11, 214, 1),
 (11, 243, 1),
 (11, 318, 1),
 (12, 10, 1),
 (12, 54, 1),
 (12, 118, 1),
 (12, 215, 1),
 (13, 206, 1),
 (14, 42, 1),
 (15, 65, 1),
 (15, 120, 1),
 (15, 241, 1),
 (15, 260, 1),
 (16, 4, 2),
 (16, 5, 1),
 (16, 9, 1),
 (16, 10, 1),
 (16, 70, 1),
 (16, 145, 1),
 (16, 178, 1),
 (16, 321, 1),
 (17, 7, 2),
 (17, 115, 1),
 (17, 131, 1),
 (17, 162, 1),
 (17, 179, 1),
 (17, 185, 1),
 (17, 195, 1),
 (17, 203, 1),
 (17, 206, 1),
 (17, 207, 1),
 (18, 82, 1),
 (18, 127, 1),
 (18, 141, 1),
 (18, 144, 1),
 (18, 221, 1),
 (19, 11, 1),
 (19, 103, 1),
 (19, 146, 1),
 (19, 203, 1),
 (19, 214, 1),
 (19, 216, 1),
 (19, 321, 1),
 (20, 7, 1),
 (20, 11, 1),
 (20, 30, 1),
 (20, 72, 1),
 (20, 88, 1),
 (20, 252, 1),
 (20, 257, 2),
 (20, 260, 1),
 (20, 267, 1),
 (21, 1, 1),
 (21, 116, 1),
 (21, 118, 1),
 (21, 119, 1),
 (21, 121, 1),
 (21, 141, 1),
 (21, 207, 2),
 (21, 235, 1),
 (22, 33, 1),
 (22, 80, 1),
 (23, 1, 10),
 (24, 90, 12),
 (25, 5, 1),
 (25, 6, 1),
 (25, 75, 1),
 (25, 91, 1),
 (25, 119, 1),
 (25, 147, 1),
 (25, 180, 1),
 (25, 241, 1),
 (26, 10, 1),
 (26, 216, 1),
 (27, 101, 1),
 (27, 117, 1),
 (27, 121, 1),
 (27, 183, 2),
 (27, 238, 1),
 (27, 246, 1),
 (27, 305, 1),
 (28, 5, 1),
 (28, 74, 1),
 (28, 321, 1),
 (29, 39, 1),
 (29, 82, 1),
 (29, 132, 1),
 (29, 195, 1),
 (29, 200, 1),
 (29, 261, 1),
 (30, 5, 1),
 (30, 73, 1),
 (30, 79, 1),
 (30, 89, 1),
 (30, 164, 1),
 (30, 191, 1),
 (30, 195, 1),
 (30, 196, 1),
 (30, 246, 1),
 (30, 321, 2),
 (31, 221, 20),
 (32, 6, 13),
 (33, 206, 1),
 (33, 218, 1),
 (33, 221, 1),
 (33, 258, 1),
 (34, 49, 1),
 (35, 1, 1),
 (35, 9, 1),
 (35, 67, 1),
 (35, 194, 1),
 (35, 322, 1),
 (36, 16, 1),
 (36, 77, 1),
 (36, 91, 1),
 (36, 120, 1),
 (36, 127, 1),
 (36, 183, 1),
 (36, 215, 1),
 (37, 55, 1),
 (37, 81, 1),
 (37, 91, 1),
 (37, 115, 1),
 (37, 145, 1),
 (37, 257, 1),
 (37, 259, 1),
 (37, 260, 1),
 (37, 271, 1),
 (37, 321, 1),
 (38, 16, 1),
 (38, 119, 1),
 (38, 126, 1),
 (38, 207, 1),
 (38, 234, 1),
 (38, 255, 1),
 (38, 278, 1),
 (39, 19, 1),
 (39, 138, 1),
 (39, 141, 1),
 (39, 189, 1),
 (39, 196, 1),
 (39, 203, 1),
 (39, 269, 1),
 (39, 293, 1),
 (40, 11, 1),
 (40, 44, 1),
 (40, 54, 1),
 (40, 132, 1),
 (40, 183, 1),
 (40, 194, 1),
 (40, 207, 1),
 (40, 215, 1),
 (40, 221, 1),
 (40, 321, 1),
 (41, 16, 1),
 (41, 61, 1),
 (41, 160, 1),
 (41, 203, 1),
 (41, 218, 1),
 (41, 233, 1),
 (41, 260, 1),
 (41, 321, 1),
 (42, 1, 1),
 (42, 16, 1),
 (42, 46, 1),
 (42, 55, 2),
 (42, 72, 1),
 (42, 83, 1),
 (42, 214, 1),
 (42, 243, 1),
 (43, 73, 1),
 (43, 91, 1),
 (43, 135, 1),
 (43, 145, 1),
 (43, 163, 1),
 (43, 191, 1),
 (43, 214, 1),
 (44, 242, 12),
 (45, 11, 1),
 (45, 17, 1),
 (45, 146, 1),
 (45, 150, 1),
 (45, 155, 1),
 (45, 162, 1),
 (45, 214, 1),
 (45, 221, 1),
 (46, 214, 11),
 (47, 1, 1),
 (47, 37, 1),
 (47, 39, 1),
 (47, 171, 1),
 (47, 203, 1),
 (47, 271, 1),
 (48, 7, 1),
 (48, 36, 1),
 (48, 91, 1),
 (48, 182, 1),
 (49, 15, 2),
 (49, 91, 1),
 (49, 154, 1),
 (49, 207, 1),
 (50, 122, 1),
 (50, 147, 1),
 (50, 157, 1),
 (50, 178, 1),
 (50, 199, 1),
 (50, 270, 1),
 (50, 321, 1),
 (51, 5, 1),
 (51, 13, 1),
 (51, 39, 1),
 (51, 54, 1),
 (51, 92, 1),
 (51, 179,

In [79]:
%%sql
select t.album_id, count(t.track_id) from  invoice_line il
JOIN track t ON il.track_id = t.track_id 
group by 1

Done.


[(1, 74),
 (2, 3),
 (3, 6),
 (4, 50),
 (5, 80),
 (6, 75),
 (7, 59),
 (8, 18),
 (9, 37),
 (10, 33),
 (11, 10),
 (13, 4),
 (14, 11),
 (15, 4),
 (16, 39),
 (17, 29),
 (18, 2),
 (19, 4),
 (20, 23),
 (21, 8),
 (22, 2),
 (23, 10),
 (24, 2),
 (26, 1),
 (27, 1),
 (30, 10),
 (33, 1),
 (34, 2),
 (35, 7),
 (36, 48),
 (37, 37),
 (38, 5),
 (39, 47),
 (40, 2),
 (41, 2),
 (42, 27),
 (43, 6),
 (44, 1),
 (46, 11),
 (47, 6),
 (48, 11),
 (49, 10),
 (50, 2),
 (54, 16),
 (55, 39),
 (56, 1),
 (57, 2),
 (58, 5),
 (59, 2),
 (61, 7),
 (62, 3),
 (63, 4),
 (64, 5),
 (65, 12),
 (66, 5),
 (67, 28),
 (69, 3),
 (70, 1),
 (72, 60),
 (73, 34),
 (74, 6),
 (75, 8),
 (76, 5),
 (77, 5),
 (78, 1),
 (79, 18),
 (80, 42),
 (81, 25),
 (82, 36),
 (83, 74),
 (84, 2),
 (85, 2),
 (86, 4),
 (87, 4),
 (88, 96),
 (89, 58),
 (90, 33),
 (91, 51),
 (92, 40),
 (93, 7),
 (94, 4),
 (95, 7),
 (96, 3),
 (97, 1),
 (98, 3),
 (99, 2),
 (100, 10),
 (101, 5),
 (102, 5),
 (103, 4),
 (104, 1),
 (105, 11),
 (106, 2),
 (107, 4),
 (108, 2),
 (109, 9),
 (110, 3),
 (111, 3),
 (112, 4),
 (113, 1),
 (115, 49),
 (116, 23),
 (117, 9),
 (118, 31),
 (119, 90),
 (120, 187),
 (121, 9),
 (122, 5),
 (123, 18),
 (125, 30),
 (126, 26),
 (127, 9),
 (128, 3),
 (129, 6),
 (130, 1),
 (131, 7),
 (132, 7),
 (133, 2),
 (134, 6),
 (135, 6),
 (136, 7),
 (137, 1),
 (138, 5),
 (139, 2),
 (140, 6),
 (141, 35),
 (142, 2),
 (143, 4),
 (144, 17),
 (145, 47),
 (146, 51),
 (147, 16),
 (148, 7),
 (149, 14),
 (150, 16),
 (151, 9),
 (152, 21),
 (153, 4),
 (154, 7),
 (155, 18),
 (156, 3),
 (157, 53),
 (158, 2),
 (159, 3),
 (160, 48),
 (162, 14),
 (163, 78),
 (164, 52),
 (166, 1),
 (169, 1),
 (170, 2),
 (171, 1),
 (173, 4),
 (174, 14),
 (175, 4),
 (176, 3),
 (178, 33),
 (179, 21),
 (180, 30),
 (181, 8),
 (182, 37),
 (183, 68),
 (184, 2),
 (185, 80),
 (186, 64),
 (188, 59),
 (189, 25),
 (190, 20),
 (191, 7),
 (192, 6),
 (193, 31),
 (194, 49),
 (195, 50),
 (196, 39),
 (197, 10),
 (198, 6),
 (199, 5),
 (200, 14),
 (203, 52),
 (204, 3),
 (206, 41),
 (207, 93),
 (209, 5),
 (210, 5),
 (211, 72),
 (212, 15),
 (213, 42),
 (214, 83),
 (215, 80),
 (216, 47),
 (217, 27),
 (218, 43),
 (221, 76),
 (222, 6),
 (223, 1),
 (224, 3),
 (225, 1),
 (227, 1),
 (230, 1),
 (232, 3),
 (233, 9),
 (234, 9),
 (235, 14),
 (236, 6),
 (237, 11),
 (238, 9),
 (239, 7),
 (240, 5),
 (241, 20),
 (242, 21),
 (243, 33),
 (244, 25),
 (245, 14),
 (246, 25),
 (247, 3),
 (248, 3),
 (252, 1),
 (255, 16),
 (256, 12),
 (257, 74),
 (258, 31),
 (259, 21),
 (260, 31),
 (261, 1),
 (262, 3),
 (264, 3),
 (265, 6),
 (266, 3),
 (267, 3),
 (268, 3),
 (269, 11),
 (270, 50),
 (271, 15),
 (277, 1),
 (278, 5),
 (279, 1),
 (280, 1),
 (293, 8),
 (302, 1),
 (304, 2),
 (305, 8),
 (314, 1),
 (318, 3),
 (321, 50),
 (322, 47),
 (323, 10),
 (325, 1),
 (329, 1),
 (336, 1),
 (339, 8),
 (344, 2)]

In [93]:
%%sql
with invoice_count as (
    select invoice_id, album_id album_id_0, count(invoice_id) bought from  invoice_line il
JOIN track t ON il.track_id = t.track_id 
group by 1,2
), album_count as (select t.album_id album_id_1, count(t.track_id) exist from  invoice_line il
JOIN track t ON il.track_id = t.track_id 
group by 1)

select (select count(*) from 
invoice_count 
join album_count on invoice_count.album_id_0 = album_count.album_id_1
where invoice_count.bought = album_count.exist) full_albums, 

(select count(*) from 
invoice_count 
join album_count on invoice_count.album_id_0 = album_count.album_id_1
where invoice_count.bought != album_count.exist) singles, count(invoice_id)
from invoice 



Done.


[(30, 3047, 614)]

# customers by mostly single tracks

In [96]:
%%sql

with invoice_count as (
    select invoice_id, album_id album_id_0, count(invoice_id) bought from  invoice_line il
JOIN track t ON il.track_id = t.track_id 
group by 1,2
), album_count as (select t.album_id album_id_1, count(t.track_id) exist from  invoice_line il
JOIN track t ON il.track_id = t.track_id 
group by 1)

select * from 
invoice_count 
join album_count on invoice_count.album_id_0 = album_count.album_id_1
where invoice_count.bought = album_count.exist

Done.


[(4, 314, 1, 314, 1),
 (16, 70, 1, 70, 1),
 (20, 252, 1, 252, 1),
 (22, 33, 1, 33, 1),
 (29, 261, 1, 261, 1),
 (40, 44, 1, 44, 1),
 (47, 171, 1, 171, 1),
 (59, 336, 1, 336, 1),
 (96, 280, 1, 280, 1),
 (111, 329, 1, 329, 1),
 (115, 26, 1, 26, 1),
 (160, 104, 1, 104, 1),
 (169, 225, 1, 225, 1),
 (180, 227, 1, 227, 1),
 (180, 277, 1, 277, 1),
 (185, 137, 1, 137, 1),
 (189, 302, 1, 302, 1),
 (195, 56, 1, 56, 1),
 (227, 166, 1, 166, 1),
 (292, 113, 1, 113, 1),
 (383, 230, 1, 230, 1),
 (421, 279, 1, 279, 1),
 (429, 78, 1, 78, 1),
 (437, 27, 1, 27, 1),
 (490, 223, 1, 223, 1),
 (567, 97, 1, 97, 1),
 (568, 169, 1, 169, 1),
 (586, 325, 1, 325, 1),
 (590, 109, 9, 109, 9),
 (600, 130, 1, 130, 1)]